instance/models/mixins/storage.py

```diff
@@ -109,3 +132,125 @@ def deprovision_swift(self):
+
+    @property
+    def bucket_name(self):
+        """
+        Return bucket name truncated to 50 characters
+        """
+        return truncate_name(
+            '{}-{}'.format(settings.AWS_S3_BUCKET_PREFIX, self.database_name.replace('_', '-')),
+            length=50
+        )
+
+    @property
+    def iam_username(self):
+        """
+        Return IAM username truncated to 50 characters
+        """
+        return truncate_name(
+            '{}-{}'.format(settings.AWS_IAM_USER_PREFIX, self.database_name),
+            length=50
+        )
+
+    def create_iam_user(self):
+        """
+        Create IAM user with access only to the s3 bucket set in s3_bucket_name
+        """
+        if not(settings.AWS_ACCESS_KEY_ID or settings.AWS_SECRET_ACCESS_KEY):
+            return
+        iam = get_master_iam_connection()
+        iam.create_user(self.iam_username)
+        iam.put_user_policy(
+            self.iam_username,
+            'allow_access_s3_bucket',
+            self.get_s3_policy()
+        )
+        key_response = iam.create_access_key(self.iam_username)
+        keys = key_response['create_access_key_response']['create_access_key_result']['access_key']
+        self.s3_access_key = keys['access_key_id']
+        self.s3_secret_access_key = keys['secret_access_key']
+
+    def get_s3_connection(self):
+        """
+        Create connection to S3 service
+        """
+        return boto.connect_s3(
+            self.s3_access_key,
+            self.s3_secret_access_key
+        )
+
+    def provision_s3(self):
+        """
+        Create S3 Bucket if it doesn't exist
+        """
+        if self.s3_access_key and self.s3_secret_access_key:
+            s3 = self.get_s3_connection()
+            bucket = s3.create_bucket(self.s3_bucket_name)
+            bucket.set_cors(get_s3_cors_config())
+
+    def deprovision_s3(self):
+        """
+        Deprovision S3 by deleting S3 bucket and IAM user
+        """
+        if not(self.s3_access_key or self.s3_secret_access_key):
+            return
+        if self.s3_bucket_name:
+            try:
+                s3 = self.get_s3_connection()
+                bucket = s3.get_bucket(self.s3_bucket_name)
+                for key in bucket:
+                    key.delete()
+                s3.delete_bucket(self.s3_bucket_name)
+                self.s3_bucket_name = ""
+                self.save()
+            except boto.exception.S3ResponseError:
+                self.logger.exception(
+                    'There was an error trying to remove S3 bucket "%s".',
+                    self.s3_bucket_name
+                )
+        try:
+            iam = get_master_iam_connection()
+            # Access keys and policies need to be deleted before removing the user
+            iam.delete_access_key(self.s3_access_key, user_name=self.iam_username)
+            iam.delete_user_policy(self.iam_username, 'allow_access_s3_bucket')
+            iam.delete_user(self.iam_username)
```

Comment:
@rocioar  The instance's IAM user isn't getting deleted, but I'm not seeing any errors either?  The inline policy and security credentials are still attached too.

Here's what I'm doing, from the shell.. Any ideas?

```python
In [1]: from instance.factories import production_instance_factory
In [2]: oc3841_3 = production_instance_factory(name="Test name", sub_domain="oc3841-3-jill.sandbox")
16/Apr/2018 11:36:59 |     INFO | process=28707 | instance.models.database_ | Creating DatabaseServer mysql-im-stage.opencraft.hosting
16/Apr/2018 11:36:59 |     INFO | process=28707 | instance.models.database_ | Creating DatabaseServer opencraft-mongodb-integration.vm.plebia.org
16/Apr/2018 11:36:59 |     INFO | process=28707 | instance.models.rabbitmq_ | Creating RabbitMQServer rabbitmq-dev.opencraft.hosting
In [3]: oc3841_3.s3_bucket_name
Out[3]: 'ocim-oc38413jill-sandbox-plebia-net'
In [4]: oc3841_3.provision_s3()
In [4]: oc3841_3.delete()
16/Apr/2018 11:38:48 |     INFO | process=28707 | instance.models.instance  | instance=150 (OC-3841 test in) | Removing New Relic Synthetics monitors
16/Apr/2018 11:38:48 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:38:52 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: preview-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:38:55 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: studio-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:38:58 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: discovery-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:39:02 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: ecommerce-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:39:18 |     INFO | process=28707 | instance.models.log_entry | Deleted 1 log entries for deleted Open edX Instance instance with ID 15
```

Reply:
@rocioar  Ok, this is really weird..  When I run the commands individually in the console, they work.  The response looks like this though:
```
{'delete_access_key_response': {'response_metadata': {'request_id': '71cfcefe-417a-11e8-9a0a-298a89e7a2b3'}}}
```
so it looks like the request is queued somewhere, which fits the [AWS Troubleshooting#Eventual consistency](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html#eventual-consistency) docs.

But if I delete the instance from the shell, even if I wait for some time, the user seems to never get deleted.  Could there be something about having to keep the IAM client alive while the request is processed?  Seems insane that that would be the case though.

Reply:
@pomegranited I've just tested this in my local environment and I'm not being able to reproduce the issue. Maybe I could connect to the host in which you are running this and debug

Reply:
@pomegranited fixed now 👍